sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/trees/TreeNode.scala

```diff
@@ -618,6 +618,165 @@ abstract class TreeNode[BaseType <: TreeNode[BaseType]] extends Product with Tre
+   * multiTransformDownWithPruning or multiTransformUpWithPruning should be used.
+   *
+   * @param rule   a function used to generate transformed alternatives for a node
+   * @param cond   a Lambda expression to prune tree traversals. If `cond.apply` returns false
+   *               on a TreeNode T, skips processing T and its subtree; otherwise, processes
+   *               T and its subtree recursively.
+   * @param ruleId is a unique Id for `rule` to prune unnecessary tree traversals. When it is
+   *               UnknownRuleId, no pruning happens. Otherwise, if `rule` (with id `ruleId`)
+   *               has been marked as in effective on a TreeNode T, skips processing T and its
+   *               subtree. Do not pass it if the rule is not purely functional and reads a
+   *               varying initial state for different invocations.
+   * @return       the stream of alternatives
+   */
+  def multiTransformWithPruning(
+      cond: TreePatternBits => Boolean,
+      ruleId: RuleId = UnknownRuleId
+    )(rule: PartialFunction[BaseType, Seq[BaseType]]): Stream[BaseType] = {
+    multiTransformDownWithPruning(cond, ruleId)(rule).map(_._1)
+  }
+
+  /**
+   * Returns alternative copies of this node where `rule` has been recursively applied to it and all
+   * of its children (pre-order).
+   *
+   * @param rule the function used to generate transformed alternatives for a node
+   * @return     the stream of alternatives
+   */
+  def multiTransformDown(rule: PartialFunction[BaseType, Seq[BaseType]]): Stream[BaseType] = {
+    multiTransformDownWithPruning(AlwaysProcess.fn, UnknownRuleId)(rule).map(_._1)
+  }
+
+  /**
+   * Returns alternative copies of this node where `rule` has been recursively applied to it and all
+   * of its children (pre-order).
+   *
+   * As it is very easy to generate enormous number of alternatives when the input tree is huge or
+   * when the rule returns large number of alternatives, this function returns the alternatives as a
+   * lazy `Stream` to be able to limit the number of alternatives generated at the caller side as
+   * needed.
+   *
+   * To indicate that the original node without any transformation is a valid alternative the rule
+   * can either:
+   * - not apply or
+   * - return an empty `Seq` or
+   * - a `Seq` that contains a node that is equal to the original node.
+   *
+   * Please note that this function always consider the original node as a valid alternative (even
+   * if the original node is not included in the returned `Seq`) if the rule can transform any of
+   * the descendants of the node. E.g. consider a simple expression:
+   *   `Add(a, b)`
+   * and a rule that returns:
+   *   `Seq(1, 2)` for `a` and
+   *   `Seq(10, 20)` for `b` and
+   *   `Seq(11, 12, 21, 22)` for `Add(a, b)` (note that the original `Add(a, b)` is not returned)
+   * then the result of `multiTransform` is:
+   *   `Seq(11, 12, 21, 22, Add(1, 10), Add(2, 10), Add(1, 20), Add(2, 20))`.
```

Comment:
This is a good example, and explains why we don't have `multiTransformUp`. Can we highlight this and remove `multiTransform` as callers should be aware that it's tramsform down?

Reply:
I think it is possible to add a `multiTransformUp()` version, although this PR doesn't contain it and I'm not sure when it would make sense to use it instead of the current top-down version.
I've removed the general `multiTransform` and kept the expilicit `multiTransformDown` versions.
I've also added a few more details to the examples.
Let me know it we need more